src/ReservoirPair.sol

```diff
@@ -148,15 +148,35 @@ abstract contract ReservoirPair is IAssetManagedPair, ReservoirERC20 {
         require(aBalance0 <= type(uint104).max && aBalance1 <= type(uint104).max, "RP: OVERFLOW");
         require(aReserve0 <= type(uint104).max && aReserve1 <= type(uint104).max, "RP: OVERFLOW");
 
-        uint32 lBlockTimestamp = uint32(_currentTime());
+        uint32 lBlockTimestamp = _currentTime();
         uint32 lTimeElapsed;
         unchecked {
             // underflow is desired
             // however in the case where no swaps happen in ~68 years (2 ** 31 seconds) the timeElapsed would underflow twice
             lTimeElapsed = lBlockTimestamp - aBlockTimestampLast;
         }
-        if (lTimeElapsed > 0 && aReserve0 != 0 && aReserve1 != 0) {
-            _updateOracle(aReserve0, aReserve1, lTimeElapsed, lBlockTimestamp);
+
+        // both balance should never be zero, but necessary to check so we don't pass 0 values into arithmetic operations
+        // on the first activity of every block, accumulate using previous instant prices and write the current instant prices
+        // we create a new sample even at the time of pair creation
+        // this implies that we do not update instant prices for subsequent mint/burn/swaps in the same block
+        // but the team has assessed that this is a small risk given the very fast block times on L2s
+        // and has decided to make the tradeoff to minimize complexity
+        if (lTimeElapsed > 0 && aBalance0 & aBalance1 != 0) {
```

Comment:
what do you think of this math trick to ensure neither balances are 0? Seems to be more gas efficient across the board based on my tests

Reply:
Shouldn't it be `aBalance0 | aBalance1 != 0`? Because if either is zero then you don't want to enter the branch?

Reply:
for your condition, if either balance0 or balance1 is not zero, that's sufficient for the statement to evaluate to true, which is not what we want. We want to enter the branch when BOTH are not zero

Reply:
actually, I just realized that both won't work. 


### For the & case 

bal0 = 5 (0b101) 
bal1 = 2  (0b010)

both balances are above zero, but will still evaluate to false 

### For the | case 

bal0 = 10 (0b1010)
bal1 = 0 (0x0)

will evaluate to true, which is not what we want. 

I'll undo the changes

Reply:
changes undone

Reply:
> For the | case
> bal0 = 10 (0b1010)
> bal1 = 0 (0x0)
> 
> will evaluate to true, which is not what we want.

This is wrong, we want |. The code before did `balance0 != 0 && balance1 != 0` which is equivalent to `balance0 | balance1 != 0`

Reply:
but... 
```
0b1111 | 0b0000  != 0
```

evaluates to true, no?? Am I delusional?

Reply:
Ahh, didn't realize the original statement was checking both were non zero rather than one was non zero. `|` can be used to detect non zero but `&` cannot be used to check for the presence of a zero.

Original:
```solidity
lTimeElapsed > 0 && aBalance0 > 0 && aBalance1 > 0
```

Optimized:
```solidity
lTimeElapsed * aBalance0 * aBalance1 > 0
```

Reply:
implemented the unchecked multiplication, it indeed does save a bit of gas

Reply:
Is it possible to have it overflow and land exactly on 0? If so what is the consequence?

Reply:
spent some time analyzing. Seems that it's not possible to overflow.

`lTimeElapsed` has max value of uint32 (limited by the type itself)
`aBalance0` and `aBalance1` are max uint104 (as checked by the require statement right above) 

So it's not possible to overflow the multiplication beyond uint256

Reply:
Awesome